include/aspect/particle/particle_handler.h

```diff
@@ -0,0 +1,258 @@
+
+        /**
+         * Default destructor.
+         */
+        ~ParticleHandler();
+
+        /**
+         * Initialize the particle world.
+         */
+        void initialize();
+
+        /**
+         * Clear all particle related data.
+         */
+        void clear();
+
+        /**
+         * Return an iterator to the first particle.
+         */
+        ParticleHandler<dim,spacedim>::particle_iterator begin() const;
+
+        /**
+         * Return an iterator to the first particle.
+         */
+        ParticleHandler<dim,spacedim>::particle_iterator begin();
+
+        /**
+         * Return an iterator past the end of the particles.
+         */
+        ParticleHandler<dim,spacedim>::particle_iterator end() const;
+
+        /**
+         * Return an iterator past the end of the particles.
+         */
+        ParticleHandler<dim,spacedim>::particle_iterator end();
+
+        /**
+         * Return a pair of particle iterators that mark the begin and end of
+         * the particles in a particular cell. The last iterator is the first
+         * particle that is not longer in the cell.
+         */
+        std::pair<ParticleHandler<dim,spacedim>::particle_iterator,ParticleHandler<dim,spacedim>::particle_iterator>
+        particle_range_in_cell(const typename parallel::distributed::Triangulation<dim>::active_cell_iterator &cell) const;
+
+        /**
+         * Return an iterator past the end of the particles.
+         */
+        std::pair<ParticleHandler<dim,spacedim>::particle_iterator,ParticleHandler<dim,spacedim>::particle_iterator>
+        particle_range_in_cell(const typename parallel::distributed::Triangulation<dim>::active_cell_iterator &cell);
+
+        /**
+         * Remove a particle pointed to by the iterator.
+         */
+        void
+        remove_particle(const typename ParticleHandler<dim,spacedim>::particle_iterator &particle);
+
+        /**
+         * Access to particles in this handler.
+         * TODO: This function eventually needs to go, to not expose internal structure.
+         * This can only be done once World no longer uses this function.
+         */
+        std::multimap<types::LevelInd, Particle<dim,spacedim> > &
+        get_particles();
+
+        /**
+         * Const access to particles in this world.
+         * TODO: This function needs to go to not expose internal structure.
+         * This can only be done once World no longer uses this function.
+         */
+        const std::multimap<types::LevelInd, Particle<dim,spacedim> > &
+        get_particles() const;
+
+        /**
+         * Return the total number of particles in the simulation. Note that
+         * this function does
+         * not compute the number of particles, because that is an expensive
+         * global MPI operation. Instead it returns the number, which is
+         * updated internally every time it might change by a call to
+         * update_n_global_particles().
```

Comment:
this note is an implementation detail. I would not talk about the motivation but instead just say "Return the total number of particles that were managed by this class the last time the update_n_global_particles() function was called. The actual number of particles may have changed since then if particles have been added or removed."